modules/tutorial-advanced-minimalist-fulcro/pages/index.adoc

```diff
@@ -0,0 +1,338 @@
+
+See the https://github.com/holyjak/minimalist-fulcro-template-backendless/blob/example/bare-essentials-routing/src/com/example/ui.cljs[full code here].
+
+=== Binding the route to the URL
+
+Reflecting the current route in the URL and changing the URL according to the route is simple. We just need to hook into the HTML 5 History API to call `change-route!` when the URL changes and to create our own routing function that also updates the URL before calling `change-route!`. Mapping from the URL to the route and vice versa just requires replacing `/` with spaces and vice versa. You can look at Fulcro RAD's `routing/html5-history` for inspiration, namely at link:{url-code-rad}routing/html5_history.cljc#L96[how it updates the URL via `.pushState`] and how it link:{url-code-rad}routing/html5_history.cljc#L143[uses `.addEventListener`] for the https://developer.mozilla.org/en-US/docs/Web/API/Window/popstate_event[`"popstate"` event] to get notified of URL changes.
+
+The simplest solution is to include https://github.com/fulcrologic/fulcro-rad[fulcro-rad] as a library and to use its `com.fulcrologic.rad.routing/route-to!` instead of `change-route!`. You also need to link:{url-code-rad-demo}src/shared/com/example/client.cljs#L58[install-route-history! during app initialization] and perhaps also {url-code-rad-demo}/src/shared/com/example/client.cljs#L32[set the route at startup] (and not only when it is changed thereafter). Remember that RAD is just an addon library to Fulcro and you can choose and pick whatever parts you want.
+
+If you do not want to include RAD or if you like to learn via video, have a look at https://youtu.be/oQpmKWBm9HE?t=867[Fulcro – Part 15: Sessions and UI Routing] that shows how to do the integration manually (though with a helper history library). You may also find useful https://github.com/aeberts/fulcro-notes-public/blob/master/fulcro-videos-pt15-sessions-and-routing.org[these time-stamped notes from the video] by Alex Eberts.
+
+NOTE: To make sure the user can reload the URL from the server, make sure that any path other than images, JS etc. falls back to the application's HTML page - https://github.com/fulcrologic/fulcro-rad-demo/blob/develop/src/shared/com/example/components/ring_middleware.clj#L48[see how fulcro-rad-demo does it].
+
+=== Loading data when a component is routed to (for the first time)
+
+Often we only want to load data when it is actually necessary, i.e. when the component that needs them is being displayed. In React it is achieved via the `componentWillMount` hook. In Fulcro we have a better solution if the component is a router target - namely its `:will-enter` component option function. (A key disadvantage of componentWillMount is "cascading loads", when a component A loads its data and renders its child B, triggering B's own load etc. In Fulcro we route to a leaf target and thus it and all intermediaries can start loading data in parallel.) Let's see how it looks.
```

Comment:
Hukka: it has a pretty long parenthesed phrase, outside regular sentences. Some kind of rewording/reorganizing would do good, I think

JH: Are you talking about "(A key disadvantage ..."? It is in the parens to show it is non-essential, a side-comment. It does not IMO distrub the reader's flow b/c it comes after a finished sentence and understanding the next sentence after it does not depend on the context of the previous one. So I do not really see it as a problem. But I could simply drop the parens, if you think that's better? Cannot think of any other meaningful change.

Reply:
That's the part I mean, yes. And perhaps simply dropping the parentheses would work, or maybe break that part to a separate paragraph. I find it very unusual to have parenthesed sentences outside main level sentences in English, and that trips my reading flow somehow a bit.